pkg/storegateway/limiter.go

```diff
@@ -0,0 +1,150 @@
+package storegateway
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/thanos/pkg/store"
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/cortexproject/cortex/pkg/storage/tsdb"
+	"github.com/cortexproject/cortex/pkg/util"
+	util_log "github.com/cortexproject/cortex/pkg/util/log"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+type limiter struct {
+	limiter *store.Limiter
+}
+
+func (c *limiter) Reserve(num uint64) error {
+	return c.ReserveWithType(num, 0)
+}
+
+func (c *limiter) ReserveWithType(num uint64, _ store.StoreDataType) error {
+	err := c.limiter.Reserve(num)
+	if err != nil {
+		return httpgrpc.Errorf(http.StatusUnprocessableEntity, err.Error())
+	}
+
+	return nil
+}
+
+type compositeLimiter struct {
+	limiters []store.BytesLimiter
+}
+
+func (c *compositeLimiter) ReserveWithType(num uint64, dataType store.StoreDataType) error {
+	for _, l := range c.limiters {
+		if err := l.ReserveWithType(num, dataType); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type tokenBucketLimiter struct {
+	podTokenBucket      *util.TokenBucket
+	userTokenBucket     *util.TokenBucket
+	requestTokenBucket  *util.TokenBucket
+	getTokensToRetrieve func(tokens uint64, dataType store.StoreDataType) int64
+	dryRun              bool
+}
+
+func (t *tokenBucketLimiter) Reserve(_ uint64) error {
+	return nil
+}
+
+func (t *tokenBucketLimiter) ReserveWithType(num uint64, dataType store.StoreDataType) error {
+	tokensToRetrieve := t.getTokensToRetrieve(num, dataType)
+
+	// check request bucket
+	retrieved := t.requestTokenBucket.Retrieve(tokensToRetrieve)
+	if retrieved {
+		t.userTokenBucket.ForceRetrieve(tokensToRetrieve)
+		t.podTokenBucket.ForceRetrieve(tokensToRetrieve)
+		return nil
+	}
+
+	// if we can't retrieve from request bucket, check shared buckets
+	retrieved = t.userTokenBucket.Retrieve(tokensToRetrieve)
+	if !retrieved {
+		// if dry run, force retrieve all tokens and return nil
```

Comment:
I might miss something. May I know why we need to force retrieve all buckets when unable to retrieve the user token bucket?

Reply:
that's only if dryrun is true, just to observe how token count in the bucket would change and at which point we would have throttled if dryrun is false

Reply:
actually let me move the dryrun block to the beginning of the function for better separation

Reply:
Could we keep the number of tokens accurate here during dry run? Imagine there is a request that will run for now + 10s. If the request is throttled at now + 1s during dry run, then the tokens will not be accurate anymore. As the request may deduct tokens from the bucket at a later time between now + 1s and now + 10s.

How about a different mechanism to implement dry run for monitoring? Maybe the rate of throttled tokens,  rate of allowed tokens and rate of total tokens?